### PR TITLE
[Fix][Ezra] Directions Docs Callback Fix

### PR DESCRIPTION
**Pull request by ezerssss**
# Please explain PR reason

Fix the Directions Callback in the Docs

Related to: https://github.com/JustFly1984/react-google-maps-api/pull/3266, https://github.com/JustFly1984/react-google-maps-api/issues/3256, https://github.com/JustFly1984/react-google-maps-api/issues/3261

### CODE REVIEW

File: packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md [hunk 1]
```diff
@@ -27,18 +27,18 @@ class Directions extends Component {
     this.onMapClick = this.onMapClick.bind(this)
   }
 
-  directionsCallback (response) {
-    console.log(response)
+  directionsCallback (result, status) {
+    console.log(result)
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md:31]: please remove logs from final PR

**Comment by JustFly1984** [comment]
Well, this looks like google had changed an underlying API, which caused old code to fail with new API version. Since Docs build is broken for 2 years, I had no ability to even test it locally, not even talking about deploying updated docs version.
I personally haven't used google maps api for any project for more than 3 years and maintaining it just for the sake of Open Source spirit.

**Comment by ezerssss** [comment]
hmmm found something weird, I basically copied the component in this pr: https://github.com/JustFly1984/react-google-maps-api/pull/3266

@JustFly1984 
@IndianBoyBR 

and it worked, no more rerenders and just called the callback function once 

**BUT**

as soon as I added the `onLoad` or the `onUnmount` functions in the `DirectionService` component, the issue exists (infinite rerenders)

https://github.com/JustFly1984/react-google-maps-api/assets/72904036/3455efda-774a-4bbc-b038-39abe282310c

**Comment by JustFly1984** [comment]
Well, you are creating new anonymous arrow function, and causing re-renders. This is ABC of react performance. This library is all about abiding props referential transparency at every step. you can't just willy-nilly pass new object to props.
Use eslint with eslint-plugin-react-perf to track those cases. you should cache any object/array/function before passing it into the prop always without exceptions across your whole application, and wrap functional components exports with memo().

**Comment by ezerssss** [comment]
Could that be one of the reasons why it was causing re-renders in the original code?

Code in the docs:
image

```
<DirectionsService
                  // required
                  options={{ // eslint-disable-line react-perf/jsx-no-new-object-as-prop
                    destination: this.state.destination,
                    origin: this.state.origin,
                    travelMode: this.state.travelMode
                  }}
                  // required
                  callback={this.directionsCallback}
                  // optional
                  onLoad={directionsService => {
                    console.log('DirectionsService onLoad directionsService: ', directionsService)
                  }}
                  // optional
                  onUnmount={directionsService => {
                    console.log('DirectionsService onUnmount directionsService: ', directionsService)
                  }}
                />
```

**Comment by JustFly1984** [comment]
yes, it is just docs with simplified examples. Everybody is expected to kind of understand referential transparency and not just blindly copy the code from docs.
// eslint-disable-line react-perf/jsx-no-new-object-as-prop kind of explains that something is odd here)

**Comment by JustFly1984** [comment]
btw you can see all examples everywhere in most of the docs around react passing objects willy-nilly.
This is for obvious reason of simplicity and reducing cognitive load.
If you got bitten with referential transparency thing in javascript once, you understand the concept once and for all, and never make same mistake again. There is no tab vs spaces argument. It is just a feature of javascript, employed by react library to cause re-renders in pure (memo) functional components. 
Unpure functional components re-renders all the time without exception anyway, even without referential transparency.
So we have to have all of them pure, and use referential transparency without exceptions.

File: packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md [hunk 2]
```diff
@@ -1,262 +1,323 @@
...
+          // required
+          zoom={2}
+          // required
+          center={center}
+          // optional
+          onClick={onMapClick}
+          // optional
+          onLoad={(map) => {
+            console.log('DirectionsRenderer onLoad map: ', map);
+          }}
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md:277]: here too!

File: packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md [hunk 3]
```diff
@@ -1,262 +1,323 @@
...
+          // optional
+          onUnmount={(map) => {
+            console.log('DirectionsRenderer onUnmount map: ', map);
+          }}
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md:281]: and here

File: packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md [hunk 4]
```diff
@@ -1,262 +1,323 @@
...
+            <DirectionsRenderer
+              // required
+              options={directionsResult}
+              // optional
+              onLoad={(directionsRenderer) => {
+                console.log(
+                  'DirectionsRenderer onLoad directionsRenderer: ',
+                  directionsRenderer
+                );
+              }}
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md:307]: and here

File: packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md [hunk 5]
```diff
@@ -1,262 +1,323 @@
...
+              // optional
+              onUnmount={(directionsRenderer) => {
+                console.log(
+                  'DirectionsRenderer onUnmount directionsRenderer: ',
+                  directionsRenderer
+                );
+              }}
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md:314]: and here

**Comment by ezerssss** [comment]
> btw you can see all examples everywhere in most of the docs around react passing objects willy-nilly. This is for obvious reason of simplicity and reducing cognitive load. If you got bitten with referential transparency thing in javascript once, you understand the concept once and for all, and never make same mistake again. There is no tab vs spaces argument. It is just a feature of javascript, employed by react library to cause re-renders in pure (memo) functional components. Unpure functional components re-renders all the time without exception anyway, even without referential transparency. So we have to have all of them pure, and use referential transparency without exceptions.

got it! i guess this is the time I kinda understand that concept. thank u so much for being patient!

File: packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md (hunk 2 above)
Comment (ezerssss) [reply packages/react-google-maps-api/src/components/directions/DirectionsRenderer.md:277]: should i remove them or transfer them o a callback function instead?

**Comment by ezerssss** [comment]
@JustFly1984 i just removed the remaining optional parameters

**Review by JustFly1984 — APPROVED** [review]

**Pull request merged**